File: apps/sdk/py-sdk/src/larkup/hub.py
import os
from typing import Any, Dict, Literal, Optional
from urllib.parse import quote

import httpx

from .client import _handle_error
from .types import (
    LarkupHubClientOptions,
    ToolDetailResponse,
    ToolListResponse,
)


class LarkupHubClient:
    """Synchronous client for the Larkup Marketplace Hub catalog."""

    def __init__(self, options: Optional[LarkupHubClientOptions] = None):
        options = options or LarkupHubClientOptions()
        base_url = options.base_url or os.getenv("LARKUP_HUB_URL", "https://hub.larkup.de")
        self.base_url = base_url.rstrip("/")
        self.api_key = options.api_key
        self._client = httpx.Client(base_url=self.base_url)

    def _get(self, path: str, **kwargs: Any) -> Dict[str, Any]:
        headers = (
            {"Authorization": f"Bearer {self.api_key}"}
            if self.api_key
            else {}
        )
        response = self._client.get(path, headers=headers, **kwargs)
        _handle_error(response)
        return response.json()

    def list_tools(
        self,
        *,
        category: Optional[
            Literal[
                "media",
                "search",
                "analytics",
                "integration",
                "embedding",
                "ai",
                "automation",
                "utility",
            ]
        ] = None,
        search: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> ToolListResponse:
        """List Marketplace tools with optional filters."""
        params = {
            key: value
            for key, value in {
                "category": category,
                "search": search,
                "page": page,
                "limit": limit,
            }.items()
            if value is not None
        }
        return ToolListResponse(**self._get("/v1/tools", params=params))

    def get_tool(self, id: str) -> ToolDetailResponse:
        """Return one Marketplace tool and its version history."""
        return ToolDetailResponse(**self._get(f"/v1/tools/{quote(id, safe='')}"))

    def close(self) -> None:
        """Close the underlying HTTP client."""
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


class AsyncLarkupHubClient:
    """Asynchronous client for the Larkup Marketplace Hub catalog."""

    def __init__(self, options: Optional[LarkupHubClientOptions] = None):
        options = options or LarkupHubClientOptions()
        base_url = options.base_url or os.getenv("LARKUP_HUB_URL", "https://hub.larkup.de")
        self.base_url = base_url.rstrip("/")
        self.api_key = options.api_key
        self._client = httpx.AsyncClient(base_url=self.base_url)

    async def _get(self, path: str, **kwargs: Any) -> Dict[str, Any]:
        headers = (
            {"Authorization": f"Bearer {self.api_key}"}
            if self.api_key
            else {}
        )
        response = await self._client.get(path, headers=headers, **kwargs)
        _handle_error(response)
        return response.json()

    async def list_tools(
        self,
        *,
        category: Optional[str] = None,
        search: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> ToolListResponse:
        """List Marketplace tools with optional filters."""
        params = {
            key: value
            for key, value in {
                "category": category,
                "search": search,
                "page": page,
                "limit": limit,
            }.items()
            if value is not None
        }
        return ToolListResponse(**await self._get("/v1/tools", params=params))

    async def get_tool(self, id: str) -> ToolDetailResponse:
        """Return one Marketplace tool and its version history."""
        return ToolDetailResponse(
            **await self._get(f"/v1/tools/{quote(id, safe='')}")
        )

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

File: apps/sdk/js-sdk/src/hub.ts
import type {
  LarkupHubClientOptions,
  ListToolsRequest,
  ToolDetailResponse,
  ToolListResponse,
} from './types';
import { LarkupApiError } from './client';

export class LarkupHubClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;

  constructor(options: LarkupHubClientOptions = {}) {
    this.baseUrl = (
      options.baseUrl ??
      process.env.LARKUP_HUB_URL ??
      'https://hub.larkup.de'
    ).replace(/\/+$/, '');
    this.apiKey = options.apiKey;
  }

  private async get<T>(path: string): Promise<T> {
    const headers = new Headers();
    if (this.apiKey) headers.set('Authorization', `Bearer ${this.apiKey}`);
    const response = await fetch(`${this.baseUrl}${path}`, { headers });
    if (!response.ok) {
      let message = response.statusText;
      try {
        const body = (await response.json()) as { error?: string };
        message = body.error || message;
      } catch {}
      throw new LarkupApiError(response.status, message);
    }
    return response.json() as Promise<T>;
  }

  /** Lists Marketplace tools from the Larkup Hub catalog. */
  listTools(request: ListToolsRequest = {}): Promise<ToolListResponse> {
    const params = new URLSearchParams();
    if (request.category) params.set('category', request.category);
    if (request.search) params.set('search', request.search);
    if (request.page !== undefined) params.set('page', String(request.page));
    if (request.limit !== undefined) params.set('limit', String(request.limit));
    const query = params.size > 0 ? `?${params}` : '';
    return this.get(`/v1/tools${query}`);
  }

  /** Returns one Marketplace tool and its version history. */
  getTool(id: string): Promise<ToolDetailResponse> {
    return this.get(`/v1/tools/${encodeURIComponent(id)}`);
  }
}
